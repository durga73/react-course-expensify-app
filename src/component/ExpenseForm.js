import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {

    constructor (props) {
        super(props);
    
        this.state = { 
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount /100 ).toString(): '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt): moment(),
            calenderFocused:false,
            errorState: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChnge = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }))
    }

    onAmountChange = (e) => {
        const amount =  e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(()=> ({ amount }));
        }
    }

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt: createdAt}));
        }
    }

    onCalanderFocusChanged = ({ focused }) => {
        this.setState(()=> ({calenderFocused: focused}))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount) {
            this.setState(()=> ({ error:'Please provide description and amount' }))
        } else {
            this.setState(()=> ({ error:'' }))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note:this.state.note
            });
        }
    }

    render() {
        return (
            <form onSubmit={this.onSubmit} className='form'>
                {this.state.error && <p className='form__error'> {this.state.error} </p>}
                <input 
                    type='text' 
                    placeholder= 'Description'
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    className='text-input'
                />
                <input 
                    type='number' 
                    placeholder= 'Amount'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    className='text-input'
                />
                <SingleDatePicker 
                    date={this.state.createdAt} 
                    onDateChange={this.onDateChange} 
                    focused={this.state.calenderFocused}
                    onFocusChange={this.onCalanderFocusChanged}
                    numberOfMonths={1}
                    isOutsideRange={(day )=> false}
                />
                <textarea 
                    placeholder= 'Add a note for your expense (optional)' 
                    value= {this.state.note}
                    onChange={this.onNoteChnge}
                    className='textarea'
                > 
                </textarea> 
                <div>
                    <button className='button'>Save Expense</button>
                </div>
            </form>
          
        )

        
    }
}