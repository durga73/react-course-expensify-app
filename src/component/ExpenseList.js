import React from "react";
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';

import ExpenseListItem from "./ExpenseListItem";

export const ExpenseList = (props) => (
    <div>
        <h1> Expenses List </h1>
        {
            props.expenses.length === 0 ? (
                <p> No Expenses</p>
            ): (
                props.expenses.map((expense ) => (
                    <ExpenseListItem key={expense.id}
                        {...expense}
                    />
                ))
            )
        }
        
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};
export default connect(mapStateToProps)(ExpenseList);
