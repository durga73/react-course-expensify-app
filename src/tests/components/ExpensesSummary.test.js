import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../component/ExpensesSummary';

test('Should correctly render Expenses Summary with 1 expense', ()=>{
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={1500}/>);
    expect(wrapper).toMatchSnapshot();
})

test('Should correctly render Expenses Summary with multiple expenses', ()=>{
    const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={61500}/>);
    expect(wrapper).toMatchSnapshot();
})