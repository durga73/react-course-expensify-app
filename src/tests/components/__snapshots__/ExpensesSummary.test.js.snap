// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render Expenses Summary with 1 expense 1`] = `
<div>
  <h1>
     Viewing 
    1
     
    expense
     totaling 
    $15.00
  </h1>
</div>
`;

exports[`Should correctly render Expenses Summary with multiple expenses 1`] = `
<div>
  <h1>
     Viewing 
    3
     
    expenses
     totaling 
    $615.00
  </h1>
</div>
`;
